src/mq/gcloud-pubsub-push/mq.ts

```diff
@@ -0,0 +1,433 @@
+
+        this.logger.info(`Enqueuing message [FedifyID: ${message.id}]`, {
+            fedifyId: message.id,
+        });
+
+        try {
+            const messageId = await this.pubSubClient
+                .topic(getFullTopic(this.pubSubClient.projectId, this.topic))
+                .publishMessage({
+                    json: message,
+                    attributes: {
+                        fedifyId: message.id,
+                    },
+                });
+
+            this.logger.info(
+                `Message [FedifyID: ${message.id}] was enqueued [PubSubID: ${messageId}]`,
+                { fedifyId: message.id, pubSubId: messageId },
+            );
+        } catch (error) {
+            this.logger.error(
+                `Failed to enqueue message [FedifyID: ${message.id}]: ${error}`,
+                { fedifyId: message.id, error },
+            );
+
+            this.events.emit(this.errorEvent, error);
+        }
+    }
+
+    /**
+     * Starts the message queue
+     *
+     * @param handler {function} Message handler
+     * @param options {MessageQueueListenOptions} Options for the listen operation
+     */
+    async listen(
+        handler: (message: any) => Promise<void> | void,
+        options: MessageQueueListenOptions = {},
+    ): Promise<void> {
+        // Set up a listener to handle messages
+        this.events.on(this.messageEvent, async (message: Message) => {
+            const fedifyId = message.attributes.fedifyId ?? 'unknown';
+
+            this.logger.info(
+                `Handling message [FedifyID: ${fedifyId}, PubSubID: ${message.id}]`,
+                { fedifyId, pubSubId: message.id },
+            );
+
+            try {
+                await handler(message.data);
+
+                message.ack();
+
+                this.logger.info(
+                    `Acknowledged message [FedifyID: ${fedifyId}, PubSubID: ${message.id}]`,
+                    { fedifyId, pubSubId: message.id },
+                );
+            } catch (error) {
+                message.nack();
+
+                this.logger.error(
+                    `Failed to handle message [FedifyID: ${fedifyId}, PubSubID: ${message.id}]: ${error}`,
+                    { fedifyId, pubSubId: message.id, error },
+                );
+
+                this.events.emit(this.errorEvent, error);
+            }
+        });
+
+        // Return a promise that resolves when the message queue is stopped
+        return await new Promise((resolve) => {
+            options.signal?.addEventListener('abort', () => {
+                this.events.removeAllListeners();
+                resolve();
+            });
+        });
+    }
+
+    /**
+     * Handles a message
+     *
+     * Ensure that the message queue is listening before calling this method
+     *
+     * @param message {Message} Message to handle
+     */
+    handleMessage(message: Message): void {
+        if (this.isListening === false) {
+            this.logger.warn(
```

Comment:
This warns, but still emits the event - I think it should probably throw here as it's not a valid interaction?

Reply:
👍 Refactored how this works in https://github.com/TryGhost/ActivityPub/pull/180/commits/e076990c4502436bb9d27a00ec9568d3986944b4